solicy-vault/sources/SolicyVault.move

```diff
@@ -0,0 +1,107 @@
+module SolicyAdminAddress::SolicyVault {
+
+  use std::signer;
+  use std::error;
+  use aptos_framework::coin;
+  use aptos_framework::coin::Coin;
+
+  
+  const EAPP_NOT_INITIALIZED: u64 = 0;
+  const EVAULT_NOT_EXISTS: u64 = 1;
+  const EINVALID_BALANCE: u64 = 2;
+  const EINVALID_VALUE: u64 = 3;
+  const EINVALID_DEDICATED_INITIALIZER: u64 = 4;
+  const EINVALID_ADMIN: u64 = 5;
+  const EINVALID_COIN: u64 = 6;
+  const EAPP_IS_PAUSED: u64 = 7;
+  
+  
+  // Struct for Admin Address and Pause Status(0/1)
+  struct AppInfo has key {
+    admin_addr: address,
+    is_paused: u8,
+  }
+
+  struct VaultInfo<phantom CoinType> has key {
+    coin: Coin<CoinType>,
+  }
+
+// Initializer Function
+  public entry fun initialize_app(initializer: &signer, admin_addr: address) {
+    let initializer_addr = signer::address_of(initializer);
+    assert!(initializer_addr == @SolicyAdminAddress, error::permission_denied(EINVALID_DEDICATED_INITIALIZER));
+    move_to<AppInfo>(initializer, AppInfo {
+        admin_addr,
+        is_paused: 0,
+    });
+  }
+
+  public entry fun deposit<CoinType>(account: &signer, amount: u64) acquires AppInfo, VaultInfo {
+
+    let app_addr = @SolicyAdminAddress;
+    assert!(exists<AppInfo>(app_addr), error::not_found(EAPP_NOT_INITIALIZED));
+
+    let account_addr = signer::address_of(account);
+    let app_info = borrow_global_mut<AppInfo>(app_addr);
+    assert!(app_info.is_paused == 0, error::permission_denied(EAPP_IS_PAUSED));
+
+    // withdraw coin from account
+    let coin: Coin<CoinType> = coin::withdraw(account, amount);
+    if (!exists<VaultInfo<CoinType>>(account_addr)) {
+      move_to<VaultInfo<CoinType>>(account, VaultInfo<CoinType> {
+        coin
+      });
+      if (!coin::is_account_registered<CoinType>(account_addr)) {
+        coin::register<CoinType>(account);
+      }
+    } else {
+      // if already deposited, then update vault_info
+      let vault_info = borrow_global_mut<VaultInfo<CoinType>>(account_addr);
+      coin::merge<CoinType>(&mut vault_info.coin, coin);
+    };
+  }
+
+  public entry fun withdraw<CoinType>(account: &signer, amount: u64) acquires VaultInfo, AppInfo {
```

Comment:
we need to ensure, the depositor cannot withdraw more than what they have deposited.
Also, rename this function to instant_withdraw